### DISCUSSION TITLE
proposed feature: in-game ship design sandbox

### DISCUSSION THREAD

**Discussion by flowers9** [endless-sky/endless-sky#9014] [Ideas]
One of the fun parts of Endless Sky is working on your ship - thinking about what changes to make to it, or considering what ship you'd like to work towards getting in the future.  However, making anything other than incremental changes to the ship you're flying isn't really possible - you can look at the stock ships for sale, and outfits for sale, but if you want to work out how to put them together you either need to buy them or copy the details out of the game and work out possibilities by hand.

I discovered that it was fairly easy to adapt the existing Shipyard and Outfitter to take a made-up player with an unlimited budget, but with access to only what the player has access to (their existing ships and anything for sale).  This let players fiddle with stuff in game to see if certain designs might work (or even be possible) without going through a hassle (either in-game or out).  And, unlike plugins that grant access to shipyards and outfitters with everything available, there're no spoilers - you're only looking at stuff you already discovered.

My attempts to code this are in  #8978 (which is working code that I use myself when playing).  Some folk disagreed with the entire concept, and some were very for it, but I think the real sticking point is that I couldn't come up with a good way to go into design mode.  The best I managed were little buttons next to the existing Shipyard and Outfitter buttons (the idea of a Blueprints panel for the Player Info ui was a nice one, but there were too many issues with it, such as it not being very obvious to new players).  The small Design buttons were obvious to new players and easy to use, but didn't look quite right.

So, does anyone have ideas about a better approach to accessing a design mode for the Shipyard and Outfitter?  And what do people think about the concept itself?

For myself, I've found that once I got access to atomic engines, it was very nice to have a way to see what the max speed and turn was for various engine combinations (I generally want the fastest ship possible with a certain minimum for turn).  There are a lot of ships you might upgrade to after your starter ship, and I've had to toss a bunch of them because I simply couldn't fit in engines, power, and cooling such that it was an upgrade - which I couldn't have found out without either attempting to build them (and failing) or copying out all the data needed to fully craft them (since turn speed depends on the mass of the ship, so you need to know the mass of everything going into the ship).

**Comment by flowers9** [comment]
It might be poor form to be the first to comment on my own thread, but here goes. ;)

I should point out that the important bits are that new players can easily find design mode, that it doesn't break immersion (so no calling it a sandbox, for instance), and that it look like it belongs to the rest of the ui.  I also think it shouldn't be too accessible, although there's certainly room for disagreement there.

That said, I thought up an approach that might work (I'll need to work up some visuals for it) - have shipyard salesfolk offer access to a ship design feature as a way of encouraging sales.  It'd only be accessible from inside a shipyard (by extending the existing bottom right buttons to the left, reducing the length of the middle panel to match the right-most and keeping the interface clean), which would offer access to both a shipyard and outfitter in a design mode, with the background tinged blue (so you know it's not the actual shipyard/outfitter).  And you'd get a spiel when you first enter a shipyard (so, right at the beginning in the default start) about the availability and why it's offered.  Any changes made there would be lost after exiting the shipyard (and the spiel would mention that, saying it's to maintain privacy).  Assuming #9002 goes through, there'd be a list of outfits in the panel of the ship you're designing which you could easily record before you leave (whether to memory, paper, or screenshot, your choice ;).

**Comment by Zitchas** [comment]
I rather like the idea of a sandbox window or mode somewhere. As I mentioned in the actual PR, I rather like the idea of having it be in the Player Info panel. Perhaps as something along the lines of an "Explore the outfit possibilities of this ship." button. I know, rather a long name for a button, but I think completely disconnecting it from a planetary outfitter both makes it more useful and reduces potential for confusion.

**Comment by flowers9** [comment]
I do see your point, but I think the designing needs of early and mid- to late-game are different.  Early on, you looking at kitting out ships you can purchase with outfits you can purchase (regardless of where you happen to get those outfits from ;), so tieing the design mode to shipyards makes sense.  Later, you're starting to run across outfits (and ships) that you can't purchase, so it makes sense to want access to it even in remote areas.

You know, there's no reason we can't have it both ways.  Early game, the players can go to shipyards to use the "Design Center" they offer, mid-game there could be an outfit for sale (by an outfitter you wouldn't run across until mid-game) that, when installed on your flagship, gives you access to the same functionality through the PlayerInfo panel (so you could use it anywhere, as long as you had it installed).  Which sounds like two PRs to me - one to introduce the design center, one to add the new outfit.

**Comment by flowers9** [comment]
Standard Shipyard:
![Screenshot from 2023-07-09 06-59-35](https://github.com/endless-sky/endless-sky/assets/11161996/8e1ac0f6-a235-4a9c-b057-75e4b1b84a25)

Design Shipyard:
![Screenshot from 2023-07-09 06-59-22](https://github.com/endless-sky/endless-sky/assets/11161996/1a62557a-ad14-466e-817b-c87642922610)

**Comment by MasterOfGrey** [comment]
I'll say the same things here that I said on #8978 :

- On the basis that this only allows access to outfits your character already knows where to purchase and/or has in cargo; this is a feature I coulda killed for a few times in the past.
  - Noting for @warp-core's sake that this is specifically how that PR operated - and that no one has suggested it do otherwise.
- This is a great QOL feature that only allows you to *plan* where to go to get things, and so does not negate the decisions around travelling, only negates the frustration of wasted travel.
  - As Hector ably noted: `There's no functionality added by this [PR] that couldn't be achieved by manually calculating everything yourself; simply allowing the game to do the calculation for you is a nice QoL feature that doesn't detract from exploration or require giving the player access to any information they don't already have.`
- It would be very nice if you could, after designing a ship, get a summary info screen for the ship like you can for your actual ships in the playerinfo panel.
  - This makes even more sense if it's accessed via the PlayerInfo panel instead of the original approach.
  - Ideally this would display with different coloured text or something so it's clear that it's a design and not an actual ship.
    - This would give you a nice way of being able to screenshot a record of what you want to do.
- Most sci-fi games I've encountered in recent years have had some kind of design feature, frequently as a necessary step prior to construction/fitting.
  - I would go so far as to suggest that we're on the verge of being the odd ones for not having a feature like this, so I definitely feel it's time.
- This sort of feature provides distinctly different advantages than a Plugin, and has the advantage of allowing a planning option for players that doesn't innately place them at risk of exposure to spoilers.

**Reply by Zitchas** [reply]
Agreed on all that.

Some additional points:
1. People have said that plug-ins already do this. They don't. There is no plug-in out there that lets me play around in an outfitter that has, say, only the NE human space outfitters and Hai outfitters. (Yes, I've had pilots find Hai space before the South, Rim, or Deep). Nor is there a place where I can design stuff with just Rim/South/Remnant outfits. There's a lot of combinations possible, virtually none of them exist in any existing plug-in that I know of. 
2. Sure, a plug-in could be made to do each of these. That involves making a separate planet for every combination of outfitters possible. And then expecting the player to find and install the plug-in, go to the plug-in system, and locate the correct planet while managing to not read the labels on any of the other planets that are, by definition, spoilers that their pilot hasn't found yet. 
3. Worse yet is expecting the player to *make* a plug-in that does this. I don't think it is reasonable to expect a player to make a list of every planet they have found, go into the datafiles to turn that into a list of every outfitter they have found so far, then create a new planet somewhere (or edit one) to have all those outfitters, then edit their pilot file to have infinite credits, then go do their design work, then remove all this, remove the extra credits, and then go track down the stuff they need to build what they designed. That's a *lot* of work. At that point, I'd expect most players to just buy the stuff, and forget about the in-game economies or impacts of infinite money. Just quickly head-canon it as having set up trade-routes from everywhere to the desired "home base" and voila!
4. As a content creator, having the ability to design outfitting for ships in-game would be quite helpful. If I could dispense with having to edit pilot files or use all-content plug-ins (or equivalents), I'd do so in a heartbeat. (In the past I have really enjoyed using https://github.com/robot-rover/endless-sky-viewer for this, but it's not fully functional yet, and hasn't been updated in a while)

**Reply by flowers9** [reply]
> 4. As a content creator, having the ability to design outfitting for ships in-game would be quite helpful.

Well, with my suggested approach, you'd still need some way for the pilot to have access to the new outfits when you're testing your content (particularly if you expect them to mainly be available through looting ships, rather than through shops).  This could be as simple as adding a new ship/outfit list to the New Boston shipyard/outfitter, but you'd need something.

**Reply by flowers9** [reply]
>It would be very nice if you could, after designing a ship, get a summary info screen for the ship like you can for your actual ships in the playerinfo panel

#9002 will do this, by putting a list of outfits in the ship sidebar pane (and it'll have a blue background to mark it as a design when you're in the design center), so you have all the came info as on the ship info panel.  You might have trouble fitting all the info on the screen on a smaller screen, though, since the layout is different than the ship info panel (more vertical).

**Comment by flowers9** [comment]
I liked my new idea so much I've submitted a PR for it, #9024.

Playing a new game with it, it was nice to have the access limited early on - when I decided I finally needed a combat capable ship, the first step was finding a shipyard so I could look at what my options were, which felt about right.
